File: ios/Configuration/InstallmentConfigurationParser.swift
//
// Copyright (c) 2026 Adyen N.V.
//
// This file is open source and available under the MIT license. See the LICENSE file for more info.
//

import Adyen

internal struct InstallmentConfigurationParser {

    private let dict: NSDictionary
    private let showInstallmentAmount: Bool

    internal init(configuration: NSDictionary, showInstallmentAmount: Bool = false) {
        self.dict = configuration
        self.showInstallmentAmount = showInstallmentAmount
    }

    var installmentConfiguration: InstallmentConfiguration? {
        var defaultOptions: InstallmentOptions?
        var cardBasedConfigurations: [CardType: InstallmentOptions] = [:]

        for (key, value) in dict {
            guard let keyString = key as? String,
                  let optionDict = value as? NSDictionary,
                  let valuesArray = (optionDict[CardKeys.Installment.values] as? [NSNumber])?.map({ UInt(truncating: $0) }) else {
                continue
            }

            let plansArray = optionDict[CardKeys.Installment.plans] as? [String] ?? []
            let includeRevolving = plansArray.contains(CardKeys.Installment.revolvingPlan)

            if keyString.lowercased() == CardKeys.Installment.defaultKey {
                // Default options for all cards
                defaultOptions = InstallmentOptions(
                    monthValues: valuesArray,
                    includesRevolving: includeRevolving
                )
            } else {
                // Card-specific options
                let cardType = CardType(rawValue: keyString)
                cardBasedConfigurations[cardType] =
                    InstallmentOptions(
                        monthValues: valuesArray,
                        includesRevolving: includeRevolving
                    )
            }
        }

        // Return configuration only if we have at least one option
        switch (defaultOptions, cardBasedConfigurations.isEmpty) {
        case let (.some(options), false):
            // Both defaultOptions and cardBasedOptions
            return InstallmentConfiguration(
                cardBasedOptions: cardBasedConfigurations,
                defaultOptions: options,
                showInstallmentAmount: showInstallmentAmount
            )
        case let (.some(options), true):
            // Only defaultOptions
            return InstallmentConfiguration(
                defaultOptions: options,
                showInstallmentAmount: showInstallmentAmount
            )
        case (.none, false):
            // Only cardBasedOptions
            return InstallmentConfiguration(
                cardBasedOptions: cardBasedConfigurations,
                showInstallmentAmount: showInstallmentAmount
            )
        case (.none, true):
            // Neither - return nil
            return nil
        }
    }

}

File: android/src/main/java/com/adyenreactnativesdk/configuration/InstallmentConfigurationParser.kt
/*
 * Copyright (c) 2026 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 */
package com.adyenreactnativesdk.configuration

import com.adyen.checkout.card.InstallmentConfiguration
import com.adyen.checkout.card.InstallmentOptions
import com.adyen.checkout.core.common.CardBrand
import com.facebook.react.bridge.ReadableMap

internal class InstallmentConfigurationParser(
  private val config: ReadableMap,
  private val showInstallmentAmount: Boolean = false,
) {
  companion object {
    private const val INSTALLMENT_VALUES_KEY = "values"
    private const val INSTALLMENT_PLANS_KEY = "plans"
    private const val INSTALLMENT_DEFAULT_KEY = "card"
    private const val INSTALLMENT_PLAN_REVOLVING = "revolving"
  }

  val installmentConfiguration: InstallmentConfiguration?
    get() {
      var defaultOptions: InstallmentOptions? = null
      val cardBasedOptions = mutableMapOf<CardBrand, InstallmentOptions>()

      val iterator = config.keySetIterator()
      while (iterator.hasNextKey()) {
        val key = iterator.nextKey()
        val optionMap = config.getMap(key) ?: continue

        val values =
          optionMap
            .getArray(INSTALLMENT_VALUES_KEY)
            ?.toArrayList()
            ?.mapNotNull {
              (it as? Number)?.toInt()
            }?.filter { it > 1 }
            ?.takeIf { it.isNotEmpty() }
            ?: continue
        val plans =
          if (optionMap.hasKey(INSTALLMENT_PLANS_KEY)) {
            optionMap.getArray(INSTALLMENT_PLANS_KEY)?.toArrayList()?.map { it.toString() } ?: emptyList()
          } else {
            emptyList()
          }
        val installmentOptions = buildInstallmentOptions(values, plans)

        if (key.equals(INSTALLMENT_DEFAULT_KEY, ignoreCase = true)) {
          // Default options for all cards
          defaultOptions = installmentOptions
        } else {
          // Card-specific options
          cardBasedOptions[CardBrand(key)] = installmentOptions
        }
      }

      // Return configuration only if we have at least one option
      return if (defaultOptions != null || cardBasedOptions.isNotEmpty()) {
        InstallmentConfiguration(
          defaultOptions = defaultOptions,
          cardBasedOptions = cardBasedOptions,
          showInstallmentAmount = showInstallmentAmount,
        )
      } else {
        null
      }
    }

  private fun buildInstallmentOptions(
    values: List<Int>,
    plans: List<String>,
  ): InstallmentOptions {
    val installmentPlans = mutableListOf(InstallmentOptions.Plan.REGULAR)
    if (plans.any { it.equals(INSTALLMENT_PLAN_REVOLVING, ignoreCase = true) }) {
      installmentPlans.add(InstallmentOptions.Plan.REVOLVING)
    }
    return InstallmentOptions(
      values = values,
      plans = installmentPlans,
      preselectedValue = null,
    )
  }
}
